File: ares-fabric/src/main/java/dev/tigr/ares/fabric/impl/modules/movement/Strafe.java
package dev.tigr.ares.fabric.impl.modules.movement;

import dev.tigr.ares.core.feature.module.Category;
import dev.tigr.ares.core.feature.module.Module;
import dev.tigr.ares.core.setting.Setting;
import dev.tigr.ares.core.setting.settings.BooleanSetting;
import dev.tigr.ares.core.setting.settings.numerical.DoubleSetting;
import dev.tigr.ares.core.setting.settings.numerical.FloatSetting;


/**
 * @author Tigermouthbear
 */
@Module.Info(name = "Strafe", description = "Increase speed and control in air", category = Category.MOVEMENT)
public class Strafe extends Module {
    private final Setting<Boolean> lowHop = register(new BooleanSetting("Low Hop", false));
    private final Setting<Double> height = register(new DoubleSetting("Height", 0.3, 0.3, 0.5)).setVisibility(lowHop::getValue);
    private final Setting<Boolean> speedBool = register(new BooleanSetting("Modify Speed", true));
    private final Setting<Float> speedVal = register(new FloatSetting("Speed", 0.32f, 0.2f, 0.6f)).setVisibility(speedBool::getValue);

    @Override
    public void onMotion() {
        if(MC.player.input.movementForward != 0 || MC.player.input.movementSideways != 0) {
            MC.player.setSprinting(true);

            if(MC.player.isOnGround() && lowHop.getValue()) MC.player.addVelocity(0, height.getValue(), 0);

            if(MC.player.isOnGround()) return;

            float speed;
            if(!speedBool.getValue())
                speed = (float) Math.sqrt(MC.player.getVelocity().x * MC.player.getVelocity().x + MC.player.getVelocity().z * MC.player.getVelocity().z);
            else speed = speedVal.getValue();

            float yaw = MC.player.yaw;
            float forward = 1;

            if(MC.player.forwardSpeed < 0) {
                yaw += 180;
                forward = -0.5f;
            } else if(MC.player.forwardSpeed > 0) forward = 0.5f;

            if(MC.player.sidewaysSpeed > 0) yaw -= 90 * forward;
            if(MC.player.sidewaysSpeed < 0) yaw += 90 * forward;

            yaw = (float) Math.toRadians(yaw);

            MC.player.setVelocity(-Math.sin(yaw) * speed, MC.player.getVelocity().y, Math.cos(yaw) * speed);
        }
    }
}

File: src/main/kotlin/org/aresclient/ares/impl/instrument/module/modules/movement/Strafe.kt
package org.aresclient.ares.impl.instrument.module.modules.movement

import org.aresclient.ares.api.instruments.Module
import org.aresclient.ares.impl.util.EntityUtil.withTransverseMovement
import kotlin.math.sqrt

object Strafe: Module(Category.MOVEMENT, "Strafe", "") {

	private val low_hop = settings
		.addBoolean("Low Hop", false)

	private val height = settings
		.addDouble("Height", 0.3)
		.setMin(0.3)
		.setMax(0.5)
		.setVisibility(low_hop::getValue)

	private val modify_speed = settings
		.addBoolean("Modify Speed", false)

	private val speed = settings
		.addDouble("Speed", 0.32)
		.setMin(0.2)
		.setMax(0.6)
		.setVisibility(modify_speed::getValue)

	private val sprint = settings
		.addBoolean("Auto Sprint", true)

	override fun onMotion() {
		val player = MC.player ?: return
		val input = player.input?.playerInput ?: return
		if (!input.forward && !input.backward && !input.left && !input.right) return

		if (sprint.value) player.setSprinting(true)
		if (player.isOnGround) {
			if (low_hop.value) player.addVelocity(0.0, height.value, 0.0)
			return
		}

		val speed =
			if (!modify_speed.value) sqrt(player.velocity.x * player.velocity.x + player.velocity.z * player.velocity.z)
			else speed.value

		player.withTransverseMovement(speed)
	}
}